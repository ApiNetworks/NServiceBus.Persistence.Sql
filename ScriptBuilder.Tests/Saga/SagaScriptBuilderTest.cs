using System.IO;
using System.Text;
using ApprovalTests;
using NServiceBus.Persistence.Sql;
using NUnit.Framework;

[TestFixture]
public class SagaScriptBuilderTest
{
    [Test]
    public void CreateWithCorrelation()
    {
        var saga = new SagaDefinition(
            name: "theSaga",
            tableSuffix: "theSaga",
            correlationProperty: new CorrelationProperty
            {
                Name = "CorrelationProperty",
                Type = CorrelationMemberType.String
            }
        );

        var builder = new StringBuilder();
        using (var writer = new StringWriter(builder))
        {
            SagaScriptBuilder.BuildCreateScript(saga, writer);
        }
        var script = builder.ToString();

        SqlValidator.Validate(script);
        Approvals.Verify(script);
    }

    [Test]
    public void CreateWithNoCorrelation()
    {
        var saga = new SagaDefinition(
            tableSuffix: "theSaga",
            name: "theSaga"
        );

        var builder = new StringBuilder();
        using (var writer = new StringWriter(builder))
        {
            SagaScriptBuilder.BuildCreateScript(saga, writer);
        }
        var script = builder.ToString();

        SqlValidator.Validate(script);
        Approvals.Verify(script);
    }

    [Test]
    public void CreateWithCorrelationAndTransitional()
    {
        var saga = new SagaDefinition(
            tableSuffix: "theSaga",
            name: "theSaga",
            correlationProperty: new CorrelationProperty
            {
                Name = "CorrelationProperty",
                Type = CorrelationMemberType.String
            },
            transitionalCorrelationProperty: new CorrelationProperty
            {
                Name = "TransitionalProperty",
                Type = CorrelationMemberType.String
            }
        );

        var builder = new StringBuilder();
        using (var writer = new StringWriter(builder))
        {
            SagaScriptBuilder.BuildCreateScript(saga, writer);
        }
        var script = builder.ToString();

        SqlValidator.Validate(script);
        Approvals.Verify(script);
    }

    [Test]
    public void BuildDropScript()
    {
        var builder = new StringBuilder();
        using (var writer = new StringWriter(builder))
        {
            var saga = new SagaDefinition(
                tableSuffix: "theSaga",
                name: "theSaga"
            );
            SagaScriptBuilder.BuildDropScript(saga, writer);
        }
        var script = builder.ToString();
        SqlValidator.Validate(script);
        Approvals.Verify(script);
    }
}