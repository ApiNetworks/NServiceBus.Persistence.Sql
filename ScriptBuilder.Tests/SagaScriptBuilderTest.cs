using System.Collections.Generic;
using System.IO;
using System.Text;
using ApprovalTests;
using NServiceBus.SqlPersistence;
using NUnit.Framework;

[TestFixture]
public class SagaScriptBuilderTest
{
    [Test]
    public void BuildCreateScript()
    {
        var sagas = new List<SagaDefinition>
        {
            new SagaDefinition
            {
                Name = "theSaga",
                MappedProperties = new List<string>
                {
                    "Property1",
                    "Property2"
                }
            }
        };

        var builder = new StringBuilder();
        using (var writer = new StringWriter(builder))
        {
            SagaScriptBuilder.BuildCreateScript("theschema", "theendpointname", sagas, s => writer);
        }
        var script = builder.ToString();

        SqlValidator.Validate(script);
        Approvals.Verify(script);
    }

    [Test]
    public void BuildDropScript()
    {
        var sagas = new List<string>
        {
            "theSaga"
        };

        var builder = new StringBuilder();
        using (var writer = new StringWriter(builder))
        {
            SagaScriptBuilder.BuildDropScript("theschema", "theendpointname", sagas, s => writer);
        }
        var script = builder.ToString();
        SqlValidator.Validate(script);
        Approvals.Verify(script);
    }
}
