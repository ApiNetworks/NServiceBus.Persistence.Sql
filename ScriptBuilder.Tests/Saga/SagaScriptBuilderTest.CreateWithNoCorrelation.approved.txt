
declare @tableName nvarchar(max) = '[' + @schema + '].[' + @endpointName + 'theSaga]';

IF NOT EXISTS
(
    SELECT *
    FROM sys.objects
    WHERE
        object_id = OBJECT_ID(@tableName) AND
        type in ('U')
)
BEGIN
DECLARE @createTable nvarchar(max);
SET @createTable = '
    CREATE TABLE ' + @tableName + '(
        [Id] [uniqueidentifier] NOT NULL PRIMARY KEY,
        [Originator] [nvarchar](255),
        [OriginalMessageId] [nvarchar](255),
        [DataXml] [xml] NOT NULL,
        [PersistenceVersion] [nvarchar](23) NOT NULL,
        [SagaTypeVersion] [nvarchar](23) NOT NULL
    )
';
exec(@createTable);
END

declare @dropIndexQuery nvarchar(max);
select @dropIndexQuery =
(
    SELECT 'DROP INDEX ' + ix.name + ' ON ' + @tableName + '; '
    FROM sysindexes ix
    WHERE
        ix.Id = (select object_id from sys.objects where name = @tableName) AND
        ix.Name IS NOT null AND
        ix.Name LIKE 'Index_Correlation_%'
);
exec sp_executesql @dropIndexQuery

declare @dropPropertiesQuery nvarchar(max);
select @dropPropertiesQuery =
(
    SELECT 'ALTER TABLE ' + @tableName  + ' DROP COLUMN ' + col.COLUMN_NAME '; '
    FROM INFORMATION_SCHEMA.COLUMNS col
    WHERE
        col.TABLE_NAME = ' + @tableName  + ' AND
        col.COLUMN_NAME LIKE 'Correlation_%'
);
exec sp_executesql @dropPropertiesQuery
